### PR TITLE
Kill ArgumentShuffleExpr

### PR DESCRIPTION
**Pull request by slavapestov**
This completes a refactoring that simplifies our call argument expressions.

In Swift 5, we used a rather complex TupleShuffleExpr to represent tuple conversions as well as call arguments containing varargs and default arguments. This included array literals, which were a degenerate case of varargs.

In https://github.com/apple/swift/pull/23591 I split up TupleShuffleExpr into DestructureTupleExpr (for tuple conversions) and ArgumentShuffleExpr (for argument lists containing default arguments and varargs).

However ArgumentShuffleExpr was still far too general. Ever since https://github.com/apple/swift-evolution/blob/master/proposals/0060-defaulted-parameter-order.md was accepted for Swift 3.0, we always evaluate arguments in the order they're written in the source, so the "shuffling" part was never taken advantage of.

Recently @pschuh implemented a way to form arrays directly in SIL by lowering an ArrayExpr, without using an ArgumentShuffleExpr. This work landed in https://github.com/apple/swift/pull/23618.

This meant that Sema could now collect varargs into a single ArrayExpr, without using an ArgumentShuffleExpr. The only remaining case where an ArgumentShuffleExpr was needed was argument lists with default arguments.

I did some preliminary refactoring in https://github.com/apple/swift/pull/23650, to ensure that all argument lists are split up into individual argument expressions when building the PreparedArguments object.

This PR builds on top of that by introducing a DefaultArgumentExpr and CallerDefaultArgumentExpr. Now Sema knows how to build argument lists containing ArrayExpr for varargs, and DefaultArgumentExpr and CallerDefaultArgumentExpr for default arguments. SILGen's ArgEmitter now knows how to directly emit delayed arguments, without going through the ArgumentShuffleExpr code path.

Note that after this change, all CallExprs now have a ParenExpr or TupleExpr as their argument. So the next step is 'flattening' the representation so that a CallExpr can hold multiple argument expressions.

### CODE REVIEW

File: include/swift/AST/Expr.h [hunk 1]
```diff
@@ -5428,7 +5339,7 @@ inline bool ApplyExpr::validateArg(Expr *e) const {
   else if (isa<BinaryExpr>(this))
     return isa<TupleExpr>(e);
   else
-    return isa<ParenExpr>(e) || isa<TupleExpr>(e) || isa<ArgumentShuffleExpr>(e);
+    return isa<ParenExpr>(e) || isa<TupleExpr>(e);
```
Comment (brentdax) [inline include/swift/AST/Expr.h]: I've been waiting for this day since I first added `ApplyExpr::validateArg()`. 🎉

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by rjmccall** [comment]
Nice!

**Comment by slavapestov** [comment]
@rjmccall FWIW, I realized I might want to bring back the ArgSpecialDest stuff... if would help tidy up enum constructor call emission with payloads.

**Comment by rjmccall** [comment]
I was wondering about that.

**Pull request merged**